components/script/dom/bindings/str.rs

```diff
@@ -445,6 +454,48 @@ fn parse_date_string(value: &str) -> Result<(u32, u32, u32), ()> {
     Ok((year_int, month_int, day_int))
 }
 
+/// https://html.spec.whatwg.org/multipage/#parse-a-week-string
+fn parse_week_string(value: &str) -> Result<(u32, u32), ()> {
+    // Step 1, 2, 3
+    let mut iterator = value.split('-');
+    let year = iterator.next().ok_or(())?;
+
+    // Step 4
+    let year_int = year.parse::<u32>().map_err(|_| ())?;
+    if year.len() < 4 || year_int == 0 {
+        return Err(());
+    }
+
+    // Step 5, 6
+    let week = iterator.next().ok_or(())?;
+    let (week_first, week_last) = week.split_at(1);
+    if week_first != "W" {
+        return Err(());
+    }
+
+    // Step 7
+    let week_int = week_last.parse::<u32>().map_err(|_| ())?;
+    if week_last.len() != 2 {
+        return Err(());
+    }
+
+    // Step 8
+    let max_week = max_week_in_year(year_int);
+
+    // Step 9
+    if week_int < 1 || week_int > max_week {
+        return Err(());
+    }
+
+    // Step 10
+    if iterator.next().is_some() || &value[(value.len() - 1)..] == "-" {
```

Comment:
The last character match here is not necessary, since `split` produces an empty string for each `-` it encounters. https://play.rust-lang.org/?gist=e50b834be9123f860f52b5762c2e07d8&version=stable